app/command/moderation/tweet.js

```diff
@@ -47,11 +48,15 @@ module.exports = class TweetCommand extends discord.Command {
 	}
 
 	async run(msg, args) {
-		let message          = args.tweet;
-		let tweetLengthCount = message.length;
+		let message = args.tweet;
 
-		if (tweetLengthCount > app.config.twitter.tweet.maxLength) {
-			return msg.reply('Dein Tweet ist um ' + (tweetLengthCount - app.config.twitter.tweet.maxLength) + ' Zeichen zu lang.');
+		let parseResult = twitterText.parseTweet(message).valid;
+		if (!parseResult.valid) {
+			if (parseResult.permillage > 1000) {
+				return msg.reply('Dein Tweet ist zu lang und hat ' + (parseResult.permillage / 10).toFixed(1) + '% der maximalen Länge!');
```

Comment:
Please show the count of characters that are over the max. tweet limit instead of the percentage. Most people would rather have a definitive count of characters that they have to remove than to guess how much characters they have to delete to reach 100%.

Reply:
If the `weightedLength` should not be used to calculate how many characters are remaining until you reach the max. length for a tweet then the implementation in the `TwitterHelper.js` is incorrect. Either way there has to be a way to get the remaining/excess characters for a tweet. If the library doesn't offer a definitive value for that you could calculate that based on the percentage. I'm not very fond of the "recommendation" that Twitter offers in the documentation for twitter-text `Adopt a progress-based UI, instead of an absolute count, to indicate remaining characters.` since that is just bad UI/UX design. You shouldn't force people to guess how much they can write.

Reply:
The problem is that what Twitter perceives as single character and what a user would perceive as single character is different. An Asian Unicode char could have a weigthed length of 2. Twitter does that because that because character sets differ in information density. So to a certain extent we would lie if we say that a text exceeds the limit by 2 chars. It could be a single Unicode character that's responsible for that. 
You are correct about the twitterhelper. But we only append ASCII and that has AFAIK always a weighted length of 1.
We could ignore the Unicode problem as a known edge case and provide the weigthed length to the user. But that wouldn't be technical correct ;-)

Reply:
About the TwitterHelper: Yeah **we** are only appending ASCII but the summary from the google calender could be something entirely different, but from the technical viewpoint the counting as 2 (or more depending on the used charset) characters is the correct way here as twitter would count them this ways and if we'd do that different it could fail to be tweeted. So this can stay the way it is.

I think we can ignore the Unicode problem since Lucy's Twitter capabilities will only ever be used by our moderators or through commands that we are defining. In the rare case that one of the moderators wants to tweet something that includes non ASCII characters AND where the character count exceeds the limits of twitter this "lying" about the exceeded length is acceptable imo.

If we want to have it 100% correct we would need to parse every unicode character in there and correlate them to the perceivable characters for a human. If there already is a library for that this would be really cool to integrate but if we would have to implement that than this isn't really worth it.